File: src/test/java/no/nav/sosialhjelp/soknad/business/service/systemdata/AdresseSystemdataTest.java
package no.nav.sosialhjelp.soknad.business.service.systemdata;

import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonAdresse;
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonAdresseValg;
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonGateAdresse;
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonMatrikkelAdresse;
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde;
import no.nav.sosialhjelp.soknad.domain.SoknadUnderArbeid;
import no.nav.sosialhjelp.soknad.person.PersonService;
import no.nav.sosialhjelp.soknad.person.domain.Bostedsadresse;
import no.nav.sosialhjelp.soknad.person.domain.Matrikkeladresse;
import no.nav.sosialhjelp.soknad.person.domain.Oppholdsadresse;
import no.nav.sosialhjelp.soknad.person.domain.Person;
import no.nav.sosialhjelp.soknad.person.domain.Vegadresse;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.extension.ExtendWith;
import org.mockito.InjectMocks;
import org.mockito.Mock;
import org.mockito.junit.jupiter.MockitoExtension;

import static java.util.Collections.emptyList;
import static no.nav.sosialhjelp.soknad.business.service.soknadservice.SoknadService.createEmptyJsonInternalSoknad;
import static org.assertj.core.api.Assertions.assertThat;
import static org.mockito.ArgumentMatchers.anyString;
import static org.mockito.Mockito.when;

@ExtendWith(MockitoExtension.class)
class AdresseSystemdataTest {

    private static final String EIER = "12345678901";
    private static final Vegadresse DEFAULT_VEGADRESSE = new Vegadresse("gateveien", 1, "A", "", "0123", "poststed", "0301", "H0101", "123456");
    private static final Vegadresse ANNEN_VEGADRESSE = new Vegadresse("en annen sti", 32, null, null, "0456", "oslo", "0302", null, null);

    @Mock
    private PersonService personService;

    @InjectMocks
    private AdresseSystemdata adresseSystemdata;

    @Test
    void skalOppdatereFolkeregistrertAdresse_vegadresse_fraPdl() {
        var soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));

        var personWithBostedsadresseVegadresse = createPersonWithBostedsadresse(new Bostedsadresse("", DEFAULT_VEGADRESSE, null));
        when(personService.hentPerson(anyString())).thenReturn(personWithBostedsadresseVegadresse);

        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid, "");

        var folkeregistrertAdresse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getFolkeregistrertAdresse();
        var bostedsadresseVegadresse = personWithBostedsadresseVegadresse.getBostedsadresse().getVegadresse();

        assertThat(folkeregistrertAdresse.getKilde()).isEqualTo(JsonKilde.SYSTEM);
        assertThat(folkeregistrertAdresse.getType()).isEqualTo(JsonAdresse.Type.GATEADRESSE);
        assertThatVegadresseIsCorrectlyConverted(bostedsadresseVegadresse, folkeregistrertAdresse);
    }

    @Test
    void skalOppdatereFolkeregistrertAdresse_matrikkeladresse_fraPdl() {
        var soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        var personWithBostedsadresseMatrikkeladresse = createPersonWithBostedsadresse(
                new Bostedsadresse(
                        "",
                        null,
                        new Matrikkeladresse(
                                "matrikkelId",
                                "postnummer",
                                "poststed",
                                "tilleggsnavn",
                                "kommunenummer",
                                "bruksenhetsnummer"
                        )
                )
        );
        when(personService.hentPerson(anyString())).thenReturn(personWithBostedsadresseMatrikkeladresse);

        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid, "");

        var folkeregistrertAdresse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getFolkeregistrertAdresse();

        assertThat(folkeregistrertAdresse.getKilde()).isEqualTo(JsonKilde.SYSTEM);
        assertThat(folkeregistrertAdresse.getType()).isEqualTo(JsonAdresse.Type.MATRIKKELADRESSE);
        var matrikkeladresse = (JsonMatrikkelAdresse) folkeregistrertAdresse;
        var bostedsadresse = personWithBostedsadresseMatrikkeladresse.getBostedsadresse().getMatrikkeladresse();
        assertThat(matrikkeladresse.getBruksnummer()).isEqualTo(bostedsadresse.getBruksenhetsnummer());
        assertThat(matrikkeladresse.getKommunenummer()).isEqualTo(bostedsadresse.getKommunenummer());
    }

    @Test
    void skalOppdatereOppholdsadresseOgPostAdresseMedMidlertidigAdresse_kontaktadresse_fraPdl() {
        var soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia()
                .withOppholdsadresse(new JsonAdresse().withAdresseValg(JsonAdresseValg.MIDLERTIDIG))
                .withPostadresse(new JsonAdresse().withAdresseValg(JsonAdresseValg.MIDLERTIDIG));

        var personWithOppholdsadresse = createPersonWithBostedsadresseOgOppholdsadresse(
                new Bostedsadresse("", DEFAULT_VEGADRESSE, null),
                new Oppholdsadresse("", ANNEN_VEGADRESSE)
        );
        when(personService.hentPerson(anyString())).thenReturn(personWithOppholdsadresse);

        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid, "");

        var folkeregistrertAdresse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getFolkeregistrertAdresse();
        var oppholdsadresse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getOppholdsadresse();
        var postadresse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getPostadresse();

        assertThat(folkeregistrertAdresse.getKilde()).isEqualTo(JsonKilde.SYSTEM);
        assertThat(oppholdsadresse.getKilde()).isEqualTo(JsonKilde.SYSTEM);
        assertThat(postadresse.getKilde()).isEqualTo(JsonKilde.SYSTEM);

        assertThat(folkeregistrertAdresse.getType()).isEqualTo(JsonAdresse.Type.GATEADRESSE);
        assertThat(oppholdsadresse.getType()).isEqualTo(JsonAdresse.Type.GATEADRESSE);
        assertThat(postadresse.getType()).isEqualTo(JsonAdresse.Type.GATEADRESSE);

        var bostedsadresseVegadresse = personWithOppholdsadresse.getBostedsadresse().getVegadresse();
        var oppholdsadresseVegadresse = personWithOppholdsadresse.getOppholdsadresse().getVegadresse();

        assertThatVegadresseIsCorrectlyConverted(bostedsadresseVegadresse, folkeregistrertAdresse);
        assertThatVegadresseIsCorrectlyConverted(oppholdsadresseVegadresse, oppholdsadresse);
        assertThatVegadresseIsCorrectlyConverted(oppholdsadresseVegadresse, postadresse);
    }

    @Test
    void skalOppdatereOppholdsadresseOgPostAdresseMedFolkeregistrertAdresse_fraPdl() {
        var soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia()
                .withOppholdsadresse(new JsonAdresse().withAdresseValg(JsonAdresseValg.FOLKEREGISTRERT))
                .withPostadresse(new JsonAdresse().withAdresseValg(JsonAdresseValg.FOLKEREGISTRERT));

        var personWithBostedsadresseVegadresse = createPersonWithBostedsadresse(new Bostedsadresse("", DEFAULT_VEGADRESSE, null));
        when(personService.hentPerson(anyString())).thenReturn(personWithBostedsadresseVegadresse);

        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid, "");

        var folkeregistrertAdresse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getFolkeregistrertAdresse();
        var oppholdsadresse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getOppholdsadresse();
        var postadresse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getPostadresse();

        assertThat(folkeregistrertAdresse.getKilde()).isEqualTo(JsonKilde.SYSTEM);
        assertThat(oppholdsadresse.getKilde()).isEqualTo(JsonKilde.SYSTEM);
        assertThat(postadresse.getKilde()).isEqualTo(JsonKilde.SYSTEM);
        assertThat(folkeregistrertAdresse)
                .isEqualTo(oppholdsadresse.withAdresseValg(null))
                .isEqualTo(postadresse);
    }

    @Test
    void skalIkkeOppdatereOppholdsadresseEllerPostAdresseDersomAdresseValgErNull_fraPdl() {
        var soknadUnderArbeid = new SoknadUnderArbeid().withJsonInternalSoknad(createEmptyJsonInternalSoknad(EIER));
        soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia()
                .withOppholdsadresse(new JsonAdresse())
                .withPostadresse(new JsonAdresse());

        var personWithBostedsadresseVegadresse = createPersonWithBostedsadresse(new Bostedsadresse("", DEFAULT_VEGADRESSE, null));
        when(personService.hentPerson(anyString())).thenReturn(personWithBostedsadresseVegadresse);

        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid, "");

        var oppholdsadresse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getOppholdsadresse();
        var postadresse = soknadUnderArbeid.getJsonInternalSoknad().getSoknad().getData().getPersonalia().getPostadresse();

        assertThat(postadresse.getAdresseValg()).isNull();
        assertThat(postadresse.getType()).isNull();
        assertThat(oppholdsadresse.getAdresseValg()).isNull();
        assertThat(oppholdsadresse.getType()).isNull();
    }

    private void assertThatVegadresseIsCorrectlyConverted(Vegadresse vegadresse, JsonAdresse jsonAdresse) {
        var gateAdresse = (JsonGateAdresse) jsonAdresse;
        assertThat(gateAdresse.getBolignummer()).isEqualTo(vegadresse.getBruksenhetsnummer());
        assertThat(gateAdresse.getGatenavn()).isEqualTo(vegadresse.getAdressenavn());
        assertThat(gateAdresse.getHusbokstav()).isEqualTo(vegadresse.getHusbokstav());
        assertThat(gateAdresse.getHusnummer()).isEqualTo(vegadresse.getHusnummer().toString());
        assertThat(gateAdresse.getKommunenummer()).isEqualTo(vegadresse.getKommunenummer());
        assertThat(gateAdresse.getLandkode()).isEqualTo("NOR");
        assertThat(gateAdresse.getPostnummer()).isEqualTo(vegadresse.getPostnummer());
        assertThat(gateAdresse.getPoststed()).isEqualTo(vegadresse.getPoststed());
    }

    private Person createPersonWithBostedsadresse(Bostedsadresse bostedsadresse) {
        return new Person("fornavn", "mellomnavn", "etternavn", EIER, "ugift", emptyList(), null, bostedsadresse, null, null);
    }

    private Person createPersonWithBostedsadresseOgOppholdsadresse(Bostedsadresse bostedsadresse, Oppholdsadresse oppholdsadresse) {
        return new Person("fornavn", "mellomnavn", "etternavn", EIER, "ugift", emptyList(), null, bostedsadresse, oppholdsadresse, null);
    }
}

File: src/test/kotlin/no/nav/sosialhjelp/soknad/personalia/adresse/AdresseSystemdataTest.kt
package no.nav.sosialhjelp.soknad.personalia.adresse

import io.mockk.every
import io.mockk.mockk
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonAdresse
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonAdresseValg
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonGateAdresse
import no.nav.sbl.soknadsosialhjelp.soknad.adresse.JsonMatrikkelAdresse
import no.nav.sbl.soknadsosialhjelp.soknad.common.JsonKilde
import no.nav.sosialhjelp.soknad.innsending.OldSoknadService.Companion.createEmptyJsonInternalSoknad
import no.nav.sosialhjelp.soknad.personalia.adresse.adresseregister.HentAdresseService
import no.nav.sosialhjelp.soknad.personalia.adresse.adresseregister.domain.KartverketMatrikkelAdresse
import no.nav.sosialhjelp.soknad.personalia.person.PersonService
import no.nav.sosialhjelp.soknad.personalia.person.domain.Bostedsadresse
import no.nav.sosialhjelp.soknad.personalia.person.domain.Matrikkeladresse
import no.nav.sosialhjelp.soknad.personalia.person.domain.Oppholdsadresse
import no.nav.sosialhjelp.soknad.personalia.person.domain.Person
import no.nav.sosialhjelp.soknad.personalia.person.domain.Vegadresse
import no.nav.sosialhjelp.soknad.repository.soknadunderarbeid.SoknadUnderArbeid
import no.nav.sosialhjelp.soknad.repository.soknadunderarbeid.SoknadUnderArbeidStatus
import org.assertj.core.api.Assertions.assertThat
import org.junit.jupiter.api.Test
import java.time.LocalDateTime

internal class AdresseSystemdataTest {

    private val personService: PersonService = mockk()
    private val hentAdresseService: HentAdresseService = mockk()
    private val adresseSystemdata = AdresseSystemdata(personService, hentAdresseService)

    @Test
    fun skalOppdatereFolkeregistrertAdresse_vegadresse_fraPdl() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        val personWithBostedsadresseVegadresse =
            createPersonWithBostedsadresse(Bostedsadresse("", DEFAULT_VEGADRESSE, null))
        every { personService.hentPerson(any()) } returns personWithBostedsadresseVegadresse

        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid)

        val folkeregistrertAdresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.folkeregistrertAdresse
        val bostedsadresseVegadresse = personWithBostedsadresseVegadresse.bostedsadresse?.vegadresse
        assertThat(folkeregistrertAdresse.kilde).isEqualTo(JsonKilde.SYSTEM)
        assertThat(folkeregistrertAdresse.type).isEqualTo(JsonAdresse.Type.GATEADRESSE)
        assertThatVegadresseIsCorrectlyConverted(bostedsadresseVegadresse, folkeregistrertAdresse)
    }

    @Test
    fun `skal oppdatere folkeregistrert matrikkeladresse - uten utfyllende matrikkeladresse`() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        val personWithBostedsadresseMatrikkeladresse = createPersonWithBostedsadresse(
            Bostedsadresse(
                "",
                null,
                Matrikkeladresse(
                    "matrikkelId",
                    "postnummer",
                    "poststed",
                    "tilleggsnavn",
                    "kommunenummer",
                    "bruksenhetsnummer"
                )
            )
        )
        every { personService.hentPerson(any()) } returns personWithBostedsadresseMatrikkeladresse
        every { hentAdresseService.hentKartverketMatrikkelAdresse(any()) } returns null

        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid)

        val folkeregistrertAdresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.folkeregistrertAdresse
        assertThat(folkeregistrertAdresse.kilde).isEqualTo(JsonKilde.SYSTEM)
        assertThat(folkeregistrertAdresse.type).isEqualTo(JsonAdresse.Type.MATRIKKELADRESSE)
        val matrikkeladresse = folkeregistrertAdresse as JsonMatrikkelAdresse
        val bostedsadresse = personWithBostedsadresseMatrikkeladresse.bostedsadresse?.matrikkeladresse
        assertThat(matrikkeladresse.kommunenummer).isEqualTo(bostedsadresse?.kommunenummer)
        assertThat(matrikkeladresse.gaardsnummer).isNull()
        assertThat(matrikkeladresse.bruksnummer).isEqualTo(bostedsadresse?.bruksenhetsnummer)
        assertThat(matrikkeladresse.festenummer).isNull()
        assertThat(matrikkeladresse.seksjonsnummer).isNull()
        assertThat(matrikkeladresse.undernummer).isNull()
    }

    @Test
    fun `skal oppdatere folkeregistrert matrikkeladresse - med utfyllende matrikkeladresse`() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        val personWithBostedsadresseMatrikkeladresse = createPersonWithBostedsadresse(
            Bostedsadresse(
                "",
                null,
                Matrikkeladresse(
                    "matrikkelId",
                    "postnummer",
                    "poststed",
                    "tilleggsnavn",
                    "kommunenummer",
                    "bruksenhetsnummer"
                )
            )
        )

        val kartverketMatrikkelAdresse = KartverketMatrikkelAdresse(
            kommunenummer = "0301",
            gaardsnummer = "gaardsnummer",
            bruksnummer = "H0101",
            festenummer = "F4",
            seksjonsunmmer = null,
            undernummer = "under1",
            bydelsnummer = "030107"
        )

        every { personService.hentPerson(any()) } returns personWithBostedsadresseMatrikkeladresse
        every { hentAdresseService.hentKartverketMatrikkelAdresse(any()) } returns kartverketMatrikkelAdresse

        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid)

        val folkeregistrertAdresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.folkeregistrertAdresse
        assertThat(folkeregistrertAdresse.kilde).isEqualTo(JsonKilde.SYSTEM)
        assertThat(folkeregistrertAdresse.type).isEqualTo(JsonAdresse.Type.MATRIKKELADRESSE)
        val matrikkeladresse = folkeregistrertAdresse as JsonMatrikkelAdresse
        assertThat(matrikkeladresse.kommunenummer).isEqualTo(kartverketMatrikkelAdresse.kommunenummer)
        assertThat(matrikkeladresse.gaardsnummer).isEqualTo(kartverketMatrikkelAdresse.gaardsnummer)
        assertThat(matrikkeladresse.bruksnummer).isEqualTo(kartverketMatrikkelAdresse.bruksnummer)
        assertThat(matrikkeladresse.festenummer).isEqualTo(kartverketMatrikkelAdresse.festenummer)
        assertThat(matrikkeladresse.seksjonsnummer).isEqualTo(kartverketMatrikkelAdresse.seksjonsunmmer)
        assertThat(matrikkeladresse.undernummer).isEqualTo(kartverketMatrikkelAdresse.undernummer)
    }

    @Test
    fun skalOppdatereOppholdsadresseOgPostAdresseMedMidlertidigAdresse_oppholdsadresse_fraPdl() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia
            .withOppholdsadresse(JsonAdresse().withAdresseValg(JsonAdresseValg.MIDLERTIDIG))
            .withPostadresse(JsonAdresse().withAdresseValg(JsonAdresseValg.MIDLERTIDIG))
        val personWithOppholdsadresse = createPersonWithBostedsadresseOgOppholdsadresse(
            Bostedsadresse("", DEFAULT_VEGADRESSE, null),
            Oppholdsadresse("", ANNEN_VEGADRESSE)
        )
        every { personService.hentPerson(any()) } returns personWithOppholdsadresse

        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid)

        val folkeregistrertAdresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.folkeregistrertAdresse
        val oppholdsadresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.oppholdsadresse
        val postadresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.postadresse
        assertThat(folkeregistrertAdresse.kilde).isEqualTo(JsonKilde.SYSTEM)
        assertThat(oppholdsadresse.kilde).isEqualTo(JsonKilde.SYSTEM)
        assertThat(postadresse.kilde).isEqualTo(JsonKilde.SYSTEM)
        assertThat(folkeregistrertAdresse.type).isEqualTo(JsonAdresse.Type.GATEADRESSE)
        assertThat(oppholdsadresse.type).isEqualTo(JsonAdresse.Type.GATEADRESSE)
        assertThat(postadresse.type).isEqualTo(JsonAdresse.Type.GATEADRESSE)
        val bostedsadresseVegadresse = personWithOppholdsadresse.bostedsadresse?.vegadresse
        val oppholdsadresseVegadresse = personWithOppholdsadresse.oppholdsadresse?.vegadresse
        assertThatVegadresseIsCorrectlyConverted(bostedsadresseVegadresse, folkeregistrertAdresse)
        assertThatVegadresseIsCorrectlyConverted(oppholdsadresseVegadresse, oppholdsadresse)
        assertThatVegadresseIsCorrectlyConverted(oppholdsadresseVegadresse, postadresse)
    }

    @Test
    fun skalOppdatereOppholdsadresseOgPostAdresseMedFolkeregistrertAdresse_fraPdl() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia
            .withOppholdsadresse(JsonAdresse().withAdresseValg(JsonAdresseValg.FOLKEREGISTRERT))
            .withPostadresse(JsonAdresse().withAdresseValg(JsonAdresseValg.FOLKEREGISTRERT))
        val personWithBostedsadresseVegadresse =
            createPersonWithBostedsadresse(Bostedsadresse("", DEFAULT_VEGADRESSE, null))
        every { personService.hentPerson(any()) } returns personWithBostedsadresseVegadresse

        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid)

        val folkeregistrertAdresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.folkeregistrertAdresse
        val oppholdsadresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.oppholdsadresse
        val postadresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.postadresse
        assertThat(folkeregistrertAdresse.kilde).isEqualTo(JsonKilde.SYSTEM)
        assertThat(oppholdsadresse.kilde).isEqualTo(JsonKilde.SYSTEM)
        assertThat(postadresse.kilde).isEqualTo(JsonKilde.SYSTEM)
        assertThat(folkeregistrertAdresse)
            .isEqualTo(oppholdsadresse.withAdresseValg(null))
            .isEqualTo(postadresse)
    }

    @Test
    fun skalIkkeOppdatereOppholdsadresseEllerPostAdresseDersomAdresseValgErNull_fraPdl() {
        val soknadUnderArbeid = createSoknadUnderArbeid()
        soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia
            .withOppholdsadresse(JsonAdresse())
            .withPostadresse(JsonAdresse())
        val personWithBostedsadresseVegadresse =
            createPersonWithBostedsadresse(Bostedsadresse("", DEFAULT_VEGADRESSE, null))
        every { personService.hentPerson(any()) } returns personWithBostedsadresseVegadresse
        adresseSystemdata.updateSystemdataIn(soknadUnderArbeid)
        val oppholdsadresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.oppholdsadresse
        val postadresse = soknadUnderArbeid.jsonInternalSoknad!!.soknad.data.personalia.postadresse
        assertThat(postadresse.adresseValg).isNull()
        assertThat(postadresse.type).isNull()
        assertThat(oppholdsadresse.adresseValg).isNull()
        assertThat(oppholdsadresse.type).isNull()
    }

    private fun assertThatVegadresseIsCorrectlyConverted(vegadresse: Vegadresse?, jsonAdresse: JsonAdresse) {
        val gateAdresse = jsonAdresse as JsonGateAdresse
        assertThat(gateAdresse.bolignummer).isEqualTo(vegadresse?.bruksenhetsnummer)
        assertThat(gateAdresse.gatenavn).isEqualTo(vegadresse?.adressenavn)
        assertThat(gateAdresse.husbokstav).isEqualTo(vegadresse?.husbokstav)
        assertThat(gateAdresse.husnummer).isEqualTo(vegadresse?.husnummer.toString())
        assertThat(gateAdresse.kommunenummer).isEqualTo(vegadresse?.kommunenummer)
        assertThat(gateAdresse.landkode).isEqualTo("NOR")
        assertThat(gateAdresse.postnummer).isEqualTo(vegadresse?.postnummer)
        assertThat(gateAdresse.poststed).isEqualTo(vegadresse?.poststed)
    }

    private fun createPersonWithBostedsadresse(bostedsadresse: Bostedsadresse): Person {
        return Person(
            fornavn = "fornavn",
            mellomnavn = "mellomnavn",
            etternavn = "etternavn",
            fnr = EIER,
            sivilstatus = "ugift",
            statsborgerskap = emptyList(),
            ektefelle = null,
            bostedsadresse = bostedsadresse,
            oppholdsadresse = null,
        )
    }

    private fun createPersonWithBostedsadresseOgOppholdsadresse(
        bostedsadresse: Bostedsadresse,
        oppholdsadresse: Oppholdsadresse
    ): Person {
        return Person(
            fornavn = "fornavn",
            mellomnavn = "mellomnavn",
            etternavn = "etternavn",
            fnr = EIER,
            sivilstatus = "ugift",
            statsborgerskap = emptyList(),
            ektefelle = null,
            bostedsadresse = bostedsadresse,
            oppholdsadresse = oppholdsadresse,
        )
    }

    companion object {
        private const val EIER = "12345678901"
        private val DEFAULT_VEGADRESSE = Vegadresse(
            adressenavn = "gateveien",
            husnummer = 1,
            husbokstav = "A",
            tilleggsnavn = "",
            postnummer = "0123",
            poststed = "poststed",
            kommunenummer = "0301",
            bruksenhetsnummer = "H0101",
            bydelsnummer = "123456"
        )
        private val ANNEN_VEGADRESSE = Vegadresse(
            adressenavn = "en annen sti",
            husnummer = 32,
            husbokstav = null,
            tilleggsnavn = null,
            postnummer = "0456",
            poststed = "oslo",
            kommunenummer = "0302",
            bruksenhetsnummer = null,
            bydelsnummer = null
        )

        private fun createSoknadUnderArbeid(): SoknadUnderArbeid {
            return SoknadUnderArbeid(
                versjon = 1L,
                behandlingsId = "BEHANDLINGSID",
                tilknyttetBehandlingsId = null,
                eier = EIER,
                jsonInternalSoknad = createEmptyJsonInternalSoknad(EIER),
                status = SoknadUnderArbeidStatus.UNDER_ARBEID,
                opprettetDato = LocalDateTime.now(),
                sistEndretDato = LocalDateTime.now()
            )
        }
    }
}
